src/date.c

```diff
@@ -1,18 +1,981 @@
+  t.tm_wday = tp.dayofweek;
+
+  len = strftime(buf, sizeof(buf), frm, &t);
+  
+  d_setlocale(&prev_locale);
+  
+  return v7_create_string(v7, buf, len, 1);
+}
+
+static int d_tptodatestr(const struct timeparts *tp, char *buf, int addtz) {
+  (void)addtz;
+
+  return sprintf(buf, "%s %s %02d %d", wday_name[tp->dayofweek],
+                                    mon_name[tp->month], tp->day, tp->year);
+}
+
+static int d_tptotimestr(const struct timeparts *tp, char *buf, int addtz) {
+  int len;
+
+  len = sprintf(buf, "%02d:%02d:%02d GMT", tp->hour, tp->min, tp->sec);
+  
+  if(addtz && g_gmtoffms != 0) {
+    len = sprintf(buf+len, "%c%02d00 (%s)", g_gmtoffms > 0? '-':'+',
+                                abs((int)g_gmtoffms/msPerHour), d_gettzname());
+  }
+  
+  return (int)strlen(buf);
+}
+
+static int d_tptostr(const struct timeparts *tp, char *buf, int addtz) {
+  int len = d_tptodatestr(tp, buf, addtz);
+  *(buf + len) = ' ';
+  return d_tptotimestr(tp, buf + len + 1, addtz) + len + 1;
+}
+
+/*+++ setXXXX & getXXXX helpers ++++*/
+static struct timeparts* d_getTP(val_t val,  struct timeparts *tp,
+                                              fbreaktime breaktimefunc) {
+  etime_t time;
+  time = v7_to_double(val);
+  breaktimefunc(&time, tp);
+  return tp;
+}
+
+/* notice: holding month in calendar format (1-12, not 0-11) */
+struct dtimepartsarr {
+  etime_t args[7];
+};
+
+enum detimepartsarr { tpyear = 0, tpmonth , tpdate , tphours, tpminutes,
+                                                  tpseconds, tpmsec, tpmax };
+
+static etime_t d_changepartoftime(const etime_t *current,
+                            struct dtimepartsarr *a, fbreaktime breaktimefunc,
+                                            fmaketime maketimefunc) {
+  /* 0 = year, 1 = month , 2 = date , 3 = hours, 
+                4 = minutes, 5 = seconds, 6 = ms */
+  struct timeparts tp;
+  unsigned long i;
+  
+  int* tp_arr[7]; /* C89 doesn't handle initialization 
+                          like x = {&tp.year, &tp.month, .... } */
+  tp_arr[0] = &tp.year; tp_arr[1] = &tp.month; tp_arr[2] = &tp.day;
+  tp_arr[3] = &tp.hour; tp_arr[4] = &tp.min; tp_arr[5] = &tp.sec;
+  tp_arr[6] = &tp.msec;
+  
+  memset(&tp, 0, sizeof(tp));
+  
+  if(breaktimefunc != NULL) {
+    breaktimefunc(current, &tp);
+  }
+  
+  for(i = 0; i < ARRAY_SIZE(tp_arr); i++) {
+    if(!isnan(a->args[i])) {
+      *tp_arr[i] = (int)a->args[i];
+    }
+  }
+
+  return maketimefunc(&tp);
+}
+
+static etime_t d_time_number_from_arr(struct v7 *v7, val_t this_obj, val_t args,
+            int start_pos, fbreaktime breaktimefunc, fmaketime makefilefunc) {
+  etime_t ret_time = INVALID_TIME;
+  long cargs;
+
+  val_t objtime = i_value_of(v7, this_obj);
+
+  if((cargs=v7_array_length(v7, args)) >= 1 && objtime != V7_TAG_NAN) {
+    int i;
+    struct dtimepartsarr a = {{INVALID_TIME, INVALID_TIME, INVALID_TIME,
+                    INVALID_TIME, INVALID_TIME, INVALID_TIME, INVALID_TIME}};
```

Comment:
paren (i.e. INVALID_TIME on second line starts at same column as INVALID_TIME of the previous line)